packages/kokkos/core/src/impl/Kokkos_Core.cpp

```diff
@@ -61,8 +61,64 @@ std::stack<std::function<void()> > finalize_hooks;
 
 namespace Kokkos { namespace Impl { namespace {
 
-int get_ctest_gpu(const char* ctest_process_count_str)
+int get_ctest_gpu(const char* local_rank_str, const char* ctest_kokkos_device_type, const char* ctest_process_count_str)
 {
+  // Make sure rank is within bounds of processes specified by CTest
+  auto process_count = std::atoi(ctest_process_count_str);
+  auto local_rank = std::atoi(local_rank_str);
+  if (local_rank >= process_count) {
+    return 0;
+  }
+
+  // Get the resource types allocated to this process
+  std::ostringstream ctest_process;
+  ctest_process << "CTEST_PROCESS_" << local_rank;
+  auto ctest_process_str = std::getenv(ctest_process.str().c_str());
```

Comment:
See my earlier note.  If you use `auto` here, the type of `ctest_process_str` will be `char*`.  However, modifying the result of `getenv` results in undefined behavior.  Thus, you should name this type `const char*` explicitly.

Reply:
Too bad there is not a way to state that you want data being pointed to to be `const` when using `auto`.  The problem is that in this case `const auto` that will just make the pointer `const`, not the underlying `char` data.  So therefore, I agree with @mhoemmen to explicitly use `const char*` to be safe.

But why would a C++ function in the `std` namespace return on non-const `char*` when the standards [warns "Modifying the string returned by getenv invokes undefined behavior"](https://en.cppreference.com/w/cpp/utility/program/getenv)?  That is pretty stupid for a standard library that is supposed to be const-correct.

Reply:
`const auto*` (or `auto const*`) will do what we want here.